### PR TITLE
Add message TTL feature

### PR DESCRIPTION
**Pull request by sirojan24**

### CODE REVIEW

File: modules/andes-core/broker/src/main/java/org/wso2/andes/configuration/enums/AndesConfiguration.java [hunk 1]
```diff
@@ -560,7 +560,12 @@
      */
     PERFORMANCE_TUNING_MESSAGE_EXPIRATION_BATCH_SIZE
             ("performanceTuning/messageExpiration/messageBatchSize", "1000", Integer.class),
-
+    /**
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/configuration/enums/AndesConfiguration.java]: we need a better explanation that this. could you ellaborate more on
1) how stuff works,
2) How that relates to configuration

Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/configuration/enums/AndesConfiguration.java]: 1 is not a safe margin. we will do a performance /long running tests and decide what a sensible value would be.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/Andes.java [hunk 2]
```diff
@@ -608,12 +608,13 @@ public long getMessageCountInDLCForQueue(String queueName, String dlcQueueName)
     /**
      * Get expired but not yet deleted messages from message store.
      *
-     * @param limit upper bound for number of messages to be returned
+     * @param lowerBoundMessageID lower bound message Id of the safe zone for delete
+     * @param queueName Queue name
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/Andes.java]: is this a storage queue name ? if so please change the variable name

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesKernelBoot.java [hunk 3]
```diff
@@ -70,6 +70,12 @@
     private static ScheduledExecutorService andesRecoveryTaskScheduler;
 
     /**
+     * Message Expiration worker task
+     */
+    private static MessageExpirationWorker messageExpirationWorker;
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesKernelBoot.java]: Done message expiration need to work during a database outage or network is partitioned ?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesMessageMetadata.java [hunk 4]
```diff
@@ -427,4 +427,20 @@ public void addProperty(String key, Object value) {
     public Object getProperty(String key) {
         return propertyMap.get(key);
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof AndesMessageMetadata)) return false;
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesMessageMetadata.java]: wrong coding standard.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/DeliveryMessageHandler.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * limitations under the License.
+ */
+
+package org.wso2.andes.kernel;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.andes.subscription.LocalSubscription;
+
+public class DeliveryMessageHandler implements DeliveryResponsibility {
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/DeliveryMessageHandler.java]: Need class comments.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessageFlusher.java [hunk 6]
```diff
@@ -101,30 +106,45 @@ public MessageFlusher() {
...
-        DeliveryRulesList.add(new MessagePurgeRule());
+        deliveryRulesList.add(new MessagePurgeRule());
     }
 
     /**
+     * Initialize the delivery filter chain
+     */
+    private void initializeDeliveryResponsibilityChain(){
+        //assign the head of the handler chain
+        deliveryResponsibilityHead = new PurgedMessageHandler();
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessageFlusher.java]: Nice!

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSConstants.java [hunk 7]
```diff
@@ -262,13 +262,13 @@
             + " ORDER BY " + MESSAGE_ID;
 
     protected static final String PS_SELECT_MESSAGE_IDS_FROM_QUEUE =
-            "SELECT " + MESSAGE_ID 
+            "SELECT " + MESSAGE_ID
             + " FROM " + METADATA_TABLE
             + " WHERE " + MESSAGE_ID + ">?"
             + " AND " + QUEUE_ID + "=?"
             + " AND " + DLC_QUEUE_ID + "=-1"
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSConstants.java]: would this work in PGSQL ( there is a problem with =-1 as I can remember)

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/ExpiredMessageHandler.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+ * limitations under the License.
+ */
+
+package org.wso2.andes.kernel;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.andes.subscription.LocalSubscription;
+
+public class ExpiredMessageHandler implements DeliveryResponsibility {
```
Comment (pumudu88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/ExpiredMessageHandler.java]: Add class comments

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java [hunk 9]
```diff
@@ -1257,6 +1269,33 @@ public void deleteMessagesFromExpiryQueue(LongArrayList messagesToRemove) throws
             close(connection, preparedStatement, RDBMSConstants.TASK_DELETING_FROM_EXPIRY_TABLE);
         }
     }
+/*
```
Comment (pumudu88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java]: Remove unwanted codes

File: modules/andes-core/broker/src/main/java/org/wso2/andes/amqp/AMQPUtils.java [hunk 10]
```diff
@@ -215,7 +215,12 @@ public static AndesMessageMetadata convertAMQMessageToAndesMetadata(AMQMessage a
         amqMetadata.writeToBuffer(0, buf);
 
         AndesMessageMetadata metadata = new AndesMessageMetadata(amqMessage.getMessageId(),underlying,true);
-        metadata.setExpirationTime(amqMessage.getExpiration());
+        //In andes message set the expiry time as the addition of arrival time with the ttl
+        long messageExpiryTime = 0L;
+        if(amqMessage.getExpiration() != 0L){
```
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/amqp/AMQPUtils.java]: The usual practice is that we have the constant on the left side of the comparison

**Pull request closed without merging**